### ISSUE TITLE
Improve Discovery

### ISSUE THREAD

**Issue by dariusc93** [Satellite-im/Warp#160]
Current implementation relies on fetching providers over DHT by default and may attempt to use a small function in a separate task for attempting to discover any peers not connected. The issue with this is that there is no direct tracking of what peers were discovered. Furthermore, with there being other methods of discovery outside of looking for peers providing a key (eg connecting through relay, searching DHT, etc), it would be best to streamline this into a single service. This will also be helpful for fetching the public key of the peer, which can be used later on (eg #159 would require that the request/response to be encrypted for said peer, in which case we wont have access to the public key from subscribed peers since gossipsub in libp2p doesnt expose the public key itself).

Thoughts:
- `Discovery` would contain a single task that would perform a check over the provider stream. Any peers found will be cached into a set in that task.
- Entry is created and a another task starts to fetch the peer identify info from libp2p/ipfs. If it succeeds, convert the public key to a did key and set it within that entry. If fails, continue in an attempt to search. 
- Manual entries can be done when performing a lookup with `DID`, if the entry does not already exist
- (Optional) Maintain connection status within the entry. 
 
Notes:
- Once the peer is discovered and/or connected, we should expect an identity push, if one is not already available locally. After a specific duration (no less than 1 minute but no more than 5 minutes) if there is no identity found to eject the peer from the discovery entry. 
- (Optional) Before starting the service so the task would run, we should add any public keys found or used prior. This would be incoming/outgoing request and friends. Discovered identities (ones found but never sent/received a request or arent friends) will not be inserted.
